aten/src/ATen/native/CPUFallback.cpp

```diff
@@ -137,62 +166,124 @@ void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack, bool
+               (input_alias_info != nullptr &&
+                *alias_info == *input_alias_info))) {
+            // We've found the original input tensor that aliases with the
+            // current output. Wrap it in an IValue and put it directly on the
+            // stack.
+            (*stack)[returns_begin + idx] = c10::IValue(tensorlist_args[i]);
+            found_alias = true;
+            break;
           }
-          // Case (2): copy case. Copy the cpu output tensor to the original device.
+        }
+      }
+      TORCH_CHECK(
+          found_alias,
+          "The operator ",
+          op.schema().operator_name(),
+          " appears to have invalid alias information. ",
+          "Found a return tensor argument with a mismatched mutable alias: ",
+          schema_returns[idx]);
+    } else {
+      if (alias_info != nullptr && !alias_info->isWrite()) {
+        // Case (3): immutable alias (view) case.
+        // Warn here, since we're copying and not creating a view.
+        // If this operator is needed, the backend should provide a kernel for
+        // it. See Note [CPU Fallback Does Not Handle View Operators]
+        std::stringstream dev_str;
+        if (tgt_device) {
+          dev_str << *tgt_device;
+        } else {
+          dev_str << "<none>";
+        }
+        if (error_on_views) {
+          TORCH_CHECK(
+              false,
+              "The operator ",
+              op.schema().operator_name(),
+              " appears to be a view operator, ",
+              "but it has no implementation for the backend \"",
+              dev_str.str(),
+              "\". View operators don't support ",
+              "since the tensor's storage cannot be shared across devices.");
+        } else {
+          TORCH_WARN(
+              false,
+              "The operator ",
+              op.schema().operator_name(),
+              " appears to be a view operator, ",
+              "but it has no implementation for the backend \"",
+              dev_str.str(),
+              "\". View operators don't support falling back to run on the CPU, ",
+              "since the tensor's storage cannot be shared across devices.");
+        }
+      }
+      // Case (2): copy case.
+      // Copy the cpu output tensor to the original device.
+
+      // We technically  might not have a target device, e.g. if you call
+      // torch.cat() with an empty list In that case, we shouldn't have any
+      // tensors to schlep across devices anyway.
+      if (tgt_device) {
+        if (returns[idx].isTensor() && returns[idx].toTensor().defined()) {
+          (*stack)[returns_begin + idx] =
+              c10::IValue(returns[idx].toTensor().to(*tgt_device));
+        } else if (
+            returns[idx].isTensorList() &&
+            validate_tensor_list(returns[idx].toTensorList())) {
+          const auto& cpu_tensors = returns[idx].toTensorList().vec();
+          std::vector<at::Tensor> tensors;
```

Comment:
nit: you might want to add:
```
tensors.reserve(cpu_tensors.size());
```

That way if you have a `_foreach_foo()` op running with a list of 500 tensors, we allocate the memory for the 500 tensor outputs up front instead of repeatedly doing more allocations as we increase the size of this output list.

Reply:
Have done.